contract/models/contract_line.py

```diff
@@ -0,0 +1,309 @@
+
+from odoo import _, api, fields, models
+from odoo.exceptions import ValidationError
+
+
+class ContractLine(models.Model):
+    _name = "contract.line"
+    _description = "Contract Line"
+    _inherit = [
+        "contract.template.line",
+        "analytic.mixin",
+    ]
+    _order = "sequence,id"
+
+    sequence = fields.Integer()
+    contract_id = fields.Many2one(
+        comodel_name="contract.contract",
+        string="Contract",
+        required=True,
+        index=True,
+        auto_join=True,
+        ondelete="cascade",
+    )
+    currency_id = fields.Many2one(related="contract_id.currency_id")
+    create_invoice_visibility = fields.Boolean(
+        compute="_compute_create_invoice_visibility"
+    )
+    active = fields.Boolean(
+        string="Active",
+        related="contract_id.active",
+        store=True,
+        readonly=True,
+    )
+
+    @api.model
+    def _compute_first_recurring_next_date(
+        self,
+        date_start,
+        recurring_invoicing_type,
+        recurring_rule_type,
+        recurring_interval,
+    ):
+        # deprecated method for backward compatibility
+        return self.get_next_invoice_date(
+            date_start,
+            recurring_invoicing_type,
+            self._get_default_recurring_invoicing_offset(
+                recurring_invoicing_type, recurring_rule_type
+            ),
+            recurring_rule_type,
+            recurring_interval,
+            max_date_end=False,
+        )
+
+    @api.constrains("recurring_next_date", "date_start")
+    def _check_recurring_next_date_start_date(self):
+        for line in self:
+            if line.display_type == "line_section" or not line.recurring_next_date:
+                continue
+            if line.date_start and line.recurring_next_date:
+                if line.date_start > line.recurring_next_date:
+                    raise ValidationError(
+                        _(
+                            "You can't have a date of next invoice anterior "
+                            "to the start of the contract line '%s'"
+                        )
+                        % line.name
+                    )
+
+    @api.constrains(
+        "date_start", "date_end", "last_date_invoiced", "recurring_next_date"
+    )
+    def _check_last_date_invoiced(self):
+        for rec in self.filtered("last_date_invoiced"):
+            if rec.date_end and rec.date_end < rec.last_date_invoiced:
+                raise ValidationError(
+                    _(
+                        "You can't have the end date before the date of last "
+                        "invoice for the contract line '%s'"
+                    )
+                    % rec.name
+                )
+            if not rec.contract_id.line_recurrence:
+                continue
+            if rec.date_start and rec.date_start > rec.last_date_invoiced:
+                raise ValidationError(
+                    _(
```

Comment:
You can replace the call to translate a string (_) with self.env._ for getting some performance improvement in some cases. See https://github.com/odoo/odoo/pull/174844.

This is only one of many occurrence of this.